### PR TITLE
Add ChatType methods to MessageChannel

### PR DESCRIPTION
**Pull request by kashike**
[**API**](https://github.com/SpongePowered/SpongeAPI/pull/1043) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/445) | [Forge](https://github.com/SpongePowered/SpongeForge/pull/523) | [Vanilla](https://github.com/SpongePowered/SpongeVanilla/pull/221)

### CODE REVIEW

File: src/main/java/org/spongepowered/api/text/channel/MessageChannel.java [hunk 1]
```diff
@@ -134,15 +137,45 @@ default void send(Text original) {
...
+     * Sends a message to this channel, transforming and sending it to
+     * the members.
+     *
+     * <p>By default, the {@link ChatTypes#SYSTEM} type is used when sending
+     * messages to applicable recipients of this channel.</p>
+     *
+     * @param original The original text to send
+     * @param type The type of message
+     */
+    default void send(Text original, ChatType type) {
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/text/channel/MessageChannel.java]: IMO. The chat type should always be the same for a MessageChannel/the caller should not be the one to controll it.

Comment (kashike) [inline src/main/java/org/spongepowered/api/text/channel/MessageChannel.java]: `send` methods should be able to control it - a message channel is a way to distribute and transform messages in bulk, there's no reason at all this should not be supported.

File: src/main/java/org/spongepowered/api/text/ChatTypeMessageReceiver.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    /**
+     * Sends the message(s) with the specified {@link ChatType} on the client.
+     *
+     * @param type The chat type to send the messages to
+     * @param messages The message(s) to send
+     */
+    default void sendMessages(ChatType type, Iterable<Text> messages) {
+        for (Text message : checkNotNull(messages, "messages")) {
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/text/ChatTypeMessageReceiver.java]: Shouldn't we check here for an empty list of messages as well? This would be more consistent with the method above.

Comment (phroa) [inline src/main/java/org/spongepowered/api/text/ChatTypeMessageReceiver.java]: Is it really an error though?

Comment (windy1) [inline src/main/java/org/spongepowered/api/text/ChatTypeMessageReceiver.java]: It won't matter anyways, in fact it's less efficient to check the size.

Comment (phroa) [inline src/main/java/org/spongepowered/api/text/ChatTypeMessageReceiver.java]: +1

Comment (kashike) [inline src/main/java/org/spongepowered/api/text/ChatTypeMessageReceiver.java]: @gabizou wanted it originally, but I'll remove it

Comment (gabizou) [inline src/main/java/org/spongepowered/api/text/ChatTypeMessageReceiver.java]: Just keep the check that the messages are not null and the type.

File: src/main/java/org/spongepowered/api/text/channel/MessageChannel.java [hunk 3]
```diff
@@ -134,15 +136,45 @@ default void send(Text original) {
...
+
+    /**
+     * Send a message to this channel, transforming and sending it to the members.
+     *
+     * @param sender The sender of the message
+     * @param original The original message to send
+     * @param type The type of message
+     */
+    default void send(@Nullable Object sender, Text original, ChatType type) {
         checkNotNull(original, "original text");
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/text/channel/MessageChannel.java]: `original` :)

**Comment by kashike** [comment]
@Minecrell could you look at merging this soon? @gabizou is attacking me for not having another PR done which depends on this :sos:

**Comment by Minecrell** [comment]
It looks fine to me. I can either merge it tomorrow or you can try to yell at someone else to do that ;)

File: src/main/java/org/spongepowered/api/text/channel/impl/DelegateMessageChannel.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import org.spongepowered.api.text.channel.MessageChannel;
+import org.spongepowered.api.text.channel.MessageReceiver;
+import org.spongepowered.api.text.chat.ChatType;
+
+import java.util.Collection;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+/**
```
Comment (windy1) [inline src/main/java/org/spongepowered/api/text/channel/impl/DelegateMessageChannel.java]: This doc is very ambiguous. What's the use case? Where is it used? Why is it necessary?

File: src/main/java/org/spongepowered/api/text/channel/impl/SimpleMutableMessageChannel.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ */
+package org.spongepowered.api.text.channel.impl;
+
+import org.spongepowered.api.text.channel.AbstractMutableMessageChannel;
+import org.spongepowered.api.text.channel.MessageReceiver;
+import org.spongepowered.api.text.channel.MutableMessageChannel;
+
+import java.util.Collection;
+
+/**
```
Comment (windy1) [inline src/main/java/org/spongepowered/api/text/channel/impl/SimpleMutableMessageChannel.java]: Same as above.

File: src/main/java/org/spongepowered/api/text/channel/AbstractMutableMessageChannel.java [hunk 6]
```diff
@@ -24,18 +24,19 @@
...
 
 /**
- * An abstract implementation of {@link MutableMessageChannel}, using a
- * {@link Set} as the underlying member list
+ * An abstract implementation of {@link MutableMessageChannel}.
  */
 public abstract class AbstractMutableMessageChannel implements MutableMessageChannel {
 
-    protected final Set<MessageReceiver> members;
+    protected final Collection<MessageReceiver> members;
```
Comment (windy1) [inline src/main/java/org/spongepowered/api/text/channel/AbstractMutableMessageChannel.java]: Why the change to Collection? Should duplicate MessageReceivers be permitted?

File: src/main/java/org/spongepowered/api/text/channel/AbstractMutableMessageChannel.java [hunk 7]
```diff
@@ -46,26 +47,28 @@ protected AbstractMutableMessageChannel() {
...
-        this.members = members;
+    protected AbstractMutableMessageChannel(Collection<MessageReceiver> members) {
+        this.members = checkNotNull(members, "members");
     }
 
     @Override
     public boolean addMember(MessageReceiver member) {
-        return this.members.add(member);
+        checkNotNull(member, "member");
+        return !this.members.contains(member) && this.members.add(member);
```
Comment (windy1) [inline src/main/java/org/spongepowered/api/text/channel/AbstractMutableMessageChannel.java]: This is why a Set was being used. Sets don't allow duplicates.

File: src/main/java/org/spongepowered/api/text/channel/AbstractMutableMessageChannel.java (hunk 6 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/channel/AbstractMutableMessageChannel.java]: It's up to the developer to avoid this - by default a `WeakHashMap`-based `Set` is used.

File: src/main/java/org/spongepowered/api/text/channel/AbstractMutableMessageChannel.java (hunk 7 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/channel/AbstractMutableMessageChannel.java]: I wrote this originally, see above comment.
